kbfscrypto/crypto_key_types.go

```diff
@@ -644,3 +669,33 @@ func UnmaskBlockCryptKey(serverHalf BlockCryptKeyServerHalf,
 	tlfCryptKey TLFCryptKey) BlockCryptKey {
 	return MakeBlockCryptKey(xorKeys(serverHalf.data, tlfCryptKey.data))
 }
+
+// BlockHashKey is used as input to encrypt/decrypt block data (v2).
+type BlockHashKey struct {
+	// Should only be used by implementations of Crypto.
+	privateByte64Container
+}
+
+// MakeBlockHashKey makes a key used for encryption and decryption for
+// the v2 block encryption scheme.
+func MakeBlockHashKey(
+	serverHalf BlockCryptKeyServerHalf, key TLFCryptKey) BlockHashKey {
+	keyData := key.Data()
+	mac := hmac.New(sha512.New, keyData[:])
+	serverHalfData := serverHalf.Data()
+	mac.Write(serverHalfData[:])
+	hash := mac.Sum(nil)
+	var hash64 [64]byte
+	copy(hash64[:], hash)
+	return BlockHashKey{privateByte64Container{hash64}}
+}
+
+func (bhk BlockHashKey) cryptKey() (key [32]byte) {
+	copy(key[:], bhk.data[:32])
```

Comment:
can you write a helper function that is like copy except it panics if both slices don't have the same length, and use it here and below? (copy just silently copies the min) I could have sworn there was already one, but I can't find it.

Reply:
But `key` is compiled to have exactly 32 bytes (in the function definition), and the second parameter will panic if it has less than 32 bytes, e.g. https://play.golang.org/p/nBHBEynQHpG.

Reply:
Hmm, you're right. I guess I'm worried about the two `32`s going out of sync in the future. Perhaps `len(key)` instead of the 2nd `32`?